src/charm.py

```diff
@@ -21,12 +21,17 @@ class JujuControllerCharm(CharmBase):
 
     def __init__(self, *args):
         super().__init__(*args)
+
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(
             self.on.dashboard_relation_joined, self._on_dashboard_relation_joined)
         self.framework.observe(
             self.on.website_relation_joined, self._on_website_relation_joined)
+        self.framework.observe(
+            self.on.dbcluster_relation_changed, self._on_dbcluster_relation_changed)
+
+        self._db_bind_address = ''
```

Comment:
it's usually a bad idea to attach state to instances like this. It may give the impression that it is persisted between runs, and it means that event handlers have to worry about _db_bind_address having been set by a previous handler, which introduces the risk for timing bugs and races

Reply:
if I understand correctly, you want to be able to tell when the ingress address has changed. 
You may need StoredState for that, or nowadays we try to always use peer relation data as controller-backed storage to survive pod churn.

Reply:
Yep, I get it. Now using `StoredState`.